source/common/router/header_formatter.cc

```diff
@@ -111,26 +119,112 @@ parseUpstreamMetadataField(const std::string& params_str) {
   };
 }
 
+// Given a string staring with `(["a", "b", "c"])`, produce a function suitable for accessing the
+// selected metadata and update the len reference to indicate how many characters were consumed.
+std::function<std::string(const Envoy::RequestInfo::RequestInfo&)>
+parseUpstreamMetadataParams(absl::string_view params, size_t& len) {
+  // Use a regex to determine where the params end. Without encoding JSON string rules for the
+  // parameters it's very difficult to distinguish the sequence ")%" within the params from the
+  // sequence ")%" at the end of the params (especially given that another %-expression may
+  // immediately follow).
+  static std::string string_match = R"EOF("(?:[^"\\[:cntrl:]]|\\["\\bfnrtu])*")EOF";
+  static std::regex params_regex(
+      R"EOF(^\(\s*\[\s*)EOF" + string_match +
+      R"EOF(\s*(?:,\s*)EOF" + string_match +
+      R"EOF(\s*)*\]\s*\))EOF");
```

Comment:
I'm not entirely happy with this part. I contend, however, that's exceedingly difficult to correctly parse certain header values without knowing whether or not a particular character is within the JSON part of the expression. For example:

`%UPSTREAM_METADATA(["X", "%%(Y)%%"])%%PROTOCOL%`
`%UPSTREAM_METADATA(["X", "Y"])%%PROTOCOL%`

Reply:
Yeah, my headache started within about 5s of looking at this. 😉 

At the very least, can you add a bunch of concrete examples in comments of what this is doing? I'm guessing that a proper recursive descent parser (or something like that) is what is needed here, but that is beyond me. @jmarantz seems to like this kind of stuff so maybe he has some additional thoughts.

Reply:
I'll think about it some more. Maybe I can write a thing that handles the non-parameterized variables and UPSTREAM_METADATA in one place.